lib/talkable/middleware.rb

```diff
@@ -25,8 +24,76 @@ def talkable_visitor_uuid(env)
       req.params[UUID] || req.cookies[UUID] || Talkable.find_or_generate_uuid
     end
 
+    def inject_uuid_in_cookie(uuid, result)
+      Rack::Utils.set_cookie_header!(result[1], UUID, {value: uuid, path: '/', expires: cookies_expiration})
+    end
+
+    def inject_uuid_in_body(uuid, result)
+      return result unless inject_body?(result)
+
+      body = result[2]
+      body_content = collect_response(body)
+      body.close if body.respond_to?(:close)
+
+      if injection_index = body_injection_position(body_content)
+        body_content = \
+          body_content[0...injection_index] \
+          << sync_uuid_content(uuid) \
+          << body_content[injection_index..-1]
+      end
+
+      if body_content
+        response = Rack::Response.new(body_content, result[0], result[1])
+        response.finish
+      else
+        result
+      end
+    end
+
     def cookies_expiration
       Time.now + (20 * 365 * 24 * 60 * 60) # 20 years
     end
+
+    def sync_uuid_url(uuid)
+      "https://www.talkable.com/public/1x1.gif?current_visitor_uuid=#{URI.escape(uuid)}"
+    end
+
+    def sync_uuid_content(uuid)
+      "\n<img src=\"#{sync_uuid_url(uuid)}\" style=\"position:absolute; left:-9999px;\" alt=\"\" />"
```

Comment:
I mean the escaping of `&amp;` not the escping of `%`

2016-08-09 14:08 GMT+03:00 Vitalik Danchenko <notifications@github.com>:

> In lib/talkable/middleware.rb
> <https://github.com/talkable/talkable-ruby/pull/4#discussion_r74038309>:
>
> >      def cookies_expiration
> >        Time.now + (20 * 365 * 24 * 60 * 60) # 20 years
> >      end
> > +
> > +    def sync_uuid_url(uuid)
> > +      "https://www.talkable.com/public/1x1.gif?current_visitor_uuid=#{URI.escape(uuid)}"
> > +    end
> > +
> > +    def sync_uuid_content(uuid)
> > +      "\n<img src=\"#{sync_uuid_url(uuid)}\" style=\"position:absolute; left:-9999px;\" alt=\"\" />"
>
> I would prefer sync_uuid_url cares about escaping, and btw furi does it.
> If we scape here as well we'll get double escaped url
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/talkable/talkable-ruby/pull/4/files/6e72826b5f81aa175a16b3ccf1835dcf544636d4#r74038309>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHeRHqNl94q-0bg9PBV9oQDfmWS3b24ks5qeF-6gaJpZM4JRwAI>
> .
>



-- 
Bogdan Gusiev.
agresso@gmail.com